### ISSUE TITLE
Techniques may fail both 1.4.1 Use of Color and 1.3.1 Info and Relationships to convey HTML element names and attributes

### ISSUE THREAD

**Issue by charleshalldesign** [w3c/wcag#4193]
All WCAG 2.2 Techniques documents that refer to HTML elements and their attributes use a convention of rendering the text in a <code> element with applied CSS to style the text differently than non code text.

However, the semantic role of <code> is not conveyed in most scenarios, leaving only the break in text elements to convey the text is different than the surrounding text – but not how.

Further, the style applied to this code text includes:
- A change from the body text color of #1d1d1d to a blue color of #036.
- This difference has a luminance contrast of only 1.3:1 which is too subtle for most to perceive and differentiate.
- A change of font from Noto Sans to Noto Sans Mono.
- This difference is too subtle for most to perceive and differentiate.

This leaves some people the possibility to perceive a visible difference and that information and relationships are being conveyed. While there is no perceivable difference when the presentation is audible.

Example:
The word ‘select’ within the first sentence of the second paragraph of the ‘Description’ section of the [H85 Technique](https://www.w3.org/WAI/WCAG22/Techniques/html/H85).

May fail 1.4.1
Because the change in font only affect some pixels of some characters (in the example, only the bottom left edge of the ‘l’ character has any perceivable change), only color is a perceivable change for the entire word.

May fail 1.3.1
Because there is an attempt to convey semantics in the audible presentation by use of <code> element, but not all instances are preceded or followed by text that conveys this by stating “element” or “attribute”.

Suggestions:

1. to better attempt to audibly convey the code element role, try using the roledescription attribute
2. to better attempt to visibly convey the difference in text, try using a different monospace font and a different color hue that has a sufficient luminance contrast difference with both the background and the body text.

**Comment by giacomo-petri** [comment]
** First Draft Working Group Response**

> May fail 1.4.1
> Because the change in font only affect some pixels of some characters (in the example, only the bottom left edge of the ‘l’ character has any perceivable change), only color is a perceivable change for the entire word.

As noted, a different font family (monospace) is used, so the distinction is not based solely on color, meaning 1.4.1 is not failing in this case. Also, the dark blue color is barely perceptible visually (even for users without low vision). Personally, I find the difference in font-family (with its added letter spacing) more helpful than color for recognizing these words.

> May fail 1.3.1
> Because there is an attempt to convey semantics in the audible presentation by use of `<code>` element, but not all instances are preceded or followed by text that conveys this by stating “element” or “attribute”.

It is programmatically identified as code, so the visual information is also available programmatically. Therefore, this does not constitute a 1.3.1 violation.


> Suggestions:
> 1. to better attempt to audibly convey the code element role, try using the roledescription attribute

I believe that, while the visual styling may offer a hint, using a roledescription that forces assistive technologies to pause and announce something uncommon could do more harm than good. I prefer maintaining a smooth, linear reading experience, allowing users to adjust their settings if they wish to have `<code>` elements announced differently.

> 2. to better attempt to visibly convey the difference in text, try using a different monospace font and a different color hue that has a sufficient luminance contrast difference with both the background and the body text.

[To discuss with the group] In my opinion, lowering the contrast with the background (which would be necessary to increase luminance contrast with the black text) is not worthwhile, since the distinction is conveyed through more than just color.

**Comment by mbgower** [comment]
@patrickhlauke to look at some explorations, and we will wait for @giacomo-petri to return before addressing

**Comment by patrickhlauke** [comment]
### Suggestion 1

Initial test using a border around existing `.hljs` large code blocks and increasing the padding slightly.

```css
padding: 1em;
border: 1px solid #909090;
```

> ![example screenshot showing the code blocks with a medium gray border](https://github.com/user-attachments/assets/03015f0e-e78f-4ed4-a145-5a6a732525ad)

### Suggestion 2a

Explored further tweaking the raw inline `code` styling to add the same border/background, and making the font size a shade smaller to give room for top/bottom padding. This might admittedly end up a bit too busy, visually.

```css
code:not(.hljs) {
    font-size: 0.85em;
    border: 1px solid #909090;
    background: #fefefe;
    padding: 0 0.25em;
    display: inline-block;
}
```

> ![example screenshot showing the inline code elements with a medium gray border](https://github.com/user-attachments/assets/b24a02e9-932a-4667-af6e-b75c44a7eb29)

**Comment by patrickhlauke** [comment]
### Suggestion 2b

a milder alternative for the inline code would be to omit border, give it a light-gray background like `#ebebeb`. similar to what github itself does here. and rationalise it by the fact that colour here isn't the ONLY visual indication that this is slightly different ... the different font does this as well. also still keeping the slightly smaller font size.

```css
code:not(.hljs) {
    font-size: 0.85em;
    background: #ebebeb;
    padding: 0 0.25em;
    display: inline-block;
}
```

> ![example of inline code, with a subtle light-gray background and smaller monospaced text](https://github.com/user-attachments/assets/059f1e21-101b-43e2-871e-ba74f3e1fe75)

**Comment by adampage** [comment]
These look terrific. I prefer [the milder alternative](https://github.com/w3c/wcag/issues/4193#issuecomment-3365584502) for the inline code elements.

**Comment by dbjorge** [comment]
I like the proposed block version.

I agree that the outlines around the inline version are uncomfortably busy and hinder readability, and that the change in font is sufficient to meet 1.4.1 without the border. I think the milder alternative looks nice, but would want to double check that the `#ebebeb` background provides sufficient text contrast in all syntax-highlighting scenarios.

**Comment by mbgower** [comment]
I also prefer the milder alternative, especially because I am trying to get designers to stroke operable controls with a 3:1 stroke, where there are not other cues. If we are doing that kind of stroke on non-operable items, it reduces our options to differentiate the two.

**Comment by baldino-m** [comment]
I personally don't really fancy the border around the block of code (Suggestion 1), since a bit too heavy. I reckon the current layout is just fine and I don't see anything wrong with it (different font-face and syntax highlighter are more than enough).  I like the "milder" option for inline code (Suggestion 2b).

**Comment by kfranqueiro** [comment]
A comment unrelated to the styles but rather related to the selectors: consider selecting on `pre` instead of `.hljs` so that the box styling also applies to `pre` elements that aren't code (e.g. [F72](https://w3c.github.io/wcag/techniques/failures/F72)), or that skip highlight.js processing via `no-highlight` (e.g. [PDF1](https://w3c.github.io/wcag/techniques/pdf/PDF1)).

RE the styles, I could give or take doing anything at all, but I'd definitely opt for the milder inline version if we do anything.

It is also worth checking how consecutive blocks will look (e.g. [ARIA19](https://w3c.github.io/wcag/techniques/aria/ARIA19)).

**Comment by stevefaulkner** [comment]
+1 to _mild_

**Comment by bruce-usab** [comment]
Discussed on Backlog call 10/31.  @giacomo-petri asked to make minor changes to his [draft response](#issuecomment-3182751321).  Stylistic concerns raised and discussed in this thread addressed (and already merged) by other PRs, including [Tweak inline code and pre-formatted blocks #4683](../pull/4683).

**Comment by giacomo-petri** [comment]
## Draft Working Group Response

After review and discussion (including the 10/31 backlog call), it was determined that the examples in question do not fail WCAG SC 1.4.1 (Use of Color) or 1.3.1 (Info and Relationships). In the first case, while color is used, the distinction is also conveyed through a change in font family (monospace) and letter spacing, providing a perceivable and programmatic distinction beyond color alone. In the second case, the `<code>` elements are programmatically identified, ensuring that their semantic meaning is conveyed to assistive technologies without reliance on visual styling.

During discussion, the group concluded that adding more visual distinction would be beneficial for overall usability and clarity. Some examples of explorations are included in the discusion in this issue. A pull request has been merged to address these visual improvements: #4683

### Rationale

> May fail 1.4.1
> Because the change in font only affect some pixels of some characters (in the example, only the bottom left edge of the ‘l’ character has any perceivable change), only color is a perceivable change for the entire word.

As noted, a different font family (`monospace`) is used, so the distinction is not based solely on color, meaning 1.4.1 is not failing in this case. Also, the dark blue color is barely perceptible visually (even for users without low vision). Personally, I find the difference in font-family (with its added letter spacing) more helpful than color for recognizing these words.

> May fail 1.3.1
> Because there is an attempt to convey semantics in the audible presentation by use of `<code>` element, but not all instances are preceded or followed by text that conveys this by stating “element” or “attribute”.

It is programmatically identified as code, so the visual information is also available programmatically. Therefore, this does not constitute a 1.3.1 violation.


> Suggestions:
> 1. to better attempt to audibly convey the code element role, try using the roledescription attribute

I believe that, while the visual styling may offer a hint, using a roledescription that forces assistive technologies to pause and announce something uncommon could do more harm than good. I prefer maintaining a smooth, linear reading experience, allowing users to adjust their settings if they wish to have `<code>` elements announced differently.

> 2. to better attempt to visibly convey the difference in text, try using a different monospace font and a different color hue that has a sufficient luminance contrast difference with both the background and the body text.

Visual enhancements already incorporated through the merged PR #4683

**Comment by mraccess77** [comment]
Relying on slight font differences alone is a very slippery slope.  Noticing or being able to detect these changes, especially while scanning paragraph or reading with only a few words in view is very difficult for people with low vision.

**Comment by mbgower** [comment]
@mraccess77 
> Relying on slight font differences alone

I agree that we want to make context clear to users. Often there is NO visual or programmatic difference when incorporating a code string in a paragraph. Whether I write aria-label or `aria-label`, the string itself is meaningful. Would you consider the first use to be a failure of any SC?

The two failures raised in the issue are a failure of Use of Color and a failure of Info and Relationships.
The inline code examples in WCAG do not rely on color alone and are programmatically indicated, so pass these 2 SCs.

I agree that we want to reinforce content differences as much as we can, but as I hope is obvious from some of the explorations above, there is also a risk of the visual presentation causing people to misinterpret what is intended.

**Comment by mbgower** [comment]
Update the draft response to the full response after the WG review period, and closing the issue.

**Comment by mraccess77** [comment]
Use of the different often monospaced font communicates to a person in the industry that that is a specific attribute/property/piece of code - so while writing it differently without using some wording like attribute, etc. would communicate something visually to the reader which could be missed by some readers with low vision. This fact might become more apparent if we used a different attribute like title and a different font which was less distinct.

In your example, the fonts are considerably different - but in other cases a font could be used that is much closer and harder to distinguish. Users might also have font overrides. While ensuring the proper markup such as the code element is used will be needed programmatically and likely satisfy WCAG - I think in some situations it may not be clear and thus an advisory could be to use fonts that are more distinct as WCAG 2.x doesn't address the situation.
